### ISSUE TITLE
JupyterLab <4 does not work on Python 3.12

### ISSUE THREAD

**Issue by naterush** [jupyterlab/jupyterlab#15332]
## Description

`pip install "jupyterlab<4"` does not work on Python 3.12.

## Reproduce

```
conda create -n python3.12-test python=3.12
conda activate python3.12-test
pip install "jupyterlab<4"
```

See that this errors, notably due to a pyzmq wheel building issue. 

## Expected behavior

JupyterLab<4 should continue to support modern versions of Python - while the Python 3 -> 4 transition is still occurring.

**Comment by andrii-i** [comment]
Thank you for opening this issue @naterush. 

Mito works with JupyterLab 4 so updating the pinned version of Mito in JupyterLab 3.6.x branch should solve the problem.

**Comment by naterush** [comment]
@andrii-i Mito actually doesn't work with JupyterLab 4 yet -- and we have a few users who are on Lab 3.x but on Python 3.12. 

Is there any where you're aware of to work around this pyzmq install issue?

Thanks!

**Comment by jtpio** [comment]
The JupyterLab CI also seems to be failing when testing on Python 3.12: https://github.com/jupyterlab/jupyterlab/pull/16009

Example run: https://github.com/jupyterlab/jupyterlab/pull/16009

Looks like an issue with the `pyzmq` package not having a wheel for that Python version?

**Comment by RNarayan73** [comment]
I have the same issue when trying to install jupyterlab 3 in a python 3.12 environment and the issue seems to be the dependency it has on jupyter-server-ydoc which requires ypy-websocket which in turn requires y-py. The last two require rust and aren't being maintained/updated for python 3.12.

@davidbrochart has made suggestions in issues https://github.com/y-crdt/ypy-websocket/issues/93#issuecomment-2021569958 and https://github.com/y-crdt/ypy/issues/152#issuecomment-1971132046 suggesting their replacement with pycrdt and pycrdt-websocket respectively, but jupyter-server-ydoc package itself just seems to have disappeared from Github! It does have a page on PyPI https://pypi.org/project/jupyter-server-ydoc/ with the last release dating to March 2023 suggesting that it has been deprecated/superseded by some other package.

Any ideas?

**Comment by davidbrochart** [comment]
> Looks like an issue with the `pyzmq` package not having a wheel for that Python version?

Yes it's installing [pyzmq v24.0.1](https://pypi.org/project/pyzmq/24.0.1/#files) which doesn't have wheels for Python 3.12.
And ypy v0.6.2 has [wheels for Python 3.12 except for Windows](https://pypi.org/project/y-py/0.6.2/#files). There was [an issue on Windows]((https://github.com/y-crdt/ypy/pull/138#issuecomment-1717219147)) at the time Python 3.12 was not officially released, but we could try again.
Anyway, we would still need wheels for pyzmq, I don't know if @minrk is planning on releasing them?

**Comment by minrk** [comment]
pyzmq's had wheels for Python 3.12 since 25.1.1 in August, before Python 3.12 was released. The problem would be in whatever package is pinning down pyzmq to a version before that, which is probably a mistake that should be fixed. Pinning down dependencies is usually not a good practice in Python, and almost always causes more problems than it solves (it caused this problem, for example).

**Comment by RNarayan73** [comment]
> There was an issue on Windows at the time Python 3.12 was not officially released, but we could try again.

@davidbrochart, that would be wonderful!

**Comment by JasonWeill** [comment]
JupyterLab 3 reached [end of maintenance](https://blog.jupyter.org/jupyterlab-3-end-of-maintenance-879778927db2) on May 15, 2024; through the end of the year, Lab 3 will only get updates for critical security issues. I strongly recommend upgrading to Lab 4, which both works with Python 3.12 and will continue to be updated.

**Comment by krassowski** [comment]
Closing as per above.
